units/dim.cpp

```diff
@@ -55,6 +42,8 @@ Dim::Dim(const std::string &label) {
     throw std::runtime_error(
         "Exceeded maximum number of different dimension labels.");
   m_id = static_cast<Id>(id);
+  read_lock.unlock();
+  const std::unique_lock write_lock(mutex);
   custom_ids[label] = m_id;
```

Comment:
This does not work when two threads want to insert two different ids at the same time.
Suppose thread 1 and 2 are both in the reading block with 2 different labels. Neither finds its label in `custom_ids` so both compute `id = size(custom_ids) + 1000` with the __same__ size.

You have to consider everything after the `custom_ids` lookup as part of the write block.

Reply:
Right, I need look write-lock earlier.

Reply:
Good catch!